File: packages/engine/python/replicate_db.py
import asyncio
import json
import sys
import os
from typing import Optional
from py_logger import PyLogger
from storage_provider import StorageProvider
from source_storage import SourceStorage
from database_client import DatabaseClient
from database_replicator import DatabaseReplicator, TableMetadata

logger = PyLogger(__name__)

async def main(
    workspace_id: str,
    source_id: str,
    dialect: Optional[str] = None,
    file_path: Optional[str] = None,
    tables_metadata: Optional[TableMetadata] = None,
) -> None:
    host = os.getenv("DATABASE_CLIENT_HOST")
    port = os.getenv("DATABASE_CLIENT_PORT")
    user = os.getenv("DATABASE_CLIENT_USER")
    password = os.getenv("DATABASE_CLIENT_PASSWORD")
    database = os.getenv("DATABASE_CLIENT_DATABASE")
    schema = os.getenv("DATABASE_CLIENT_SCHEMA")

    storage = StorageProvider(
        tmp_dir="./tmp",
    )
    
    if file_path:
        tmp_file_path = storage.download_tmp_file(file_path)

    client = DatabaseClient(
        dialect=dialect,
        host=host,
        port=port,
        user=user,
        password=password,
        database=database,
        schema=schema,
        file_path=tmp_file_path,
        tmp_dir="./tmp",
    )

    source = SourceStorage(
        workspace_id=workspace_id,
        source_id=source_id,
        storage=storage,
    )
    replicator = DatabaseReplicator(
        source=source,
        client=client,
    )

    replicator.export_tables_to_parquet(tables_metadata)

    client.cleanup_tmp_path() # Close database client
    storage.cleanup_tmp_path() # Clean up tmp path

    logger.warning({
        "status": "success",
        "message": "Database replicated successfully",
    })

if __name__ == "__main__":
    try:
        if len(sys.argv) < 2:
            logger.error({"status": "error", "message": "No args provided"})
            sys.exit(1)
        scriptArgs = json.loads(sys.argv[1])
        logger.info({ "status": None, "message": "Script started!", "args": scriptArgs })
        asyncio.run(main(**scriptArgs))
        logger.info({"status": "success", "message": "Script ended!"})
        sys.exit(0)
    except Exception as e:
        logger.error({"status": "error", "message": str(e)})
        sys.exit(1)

File: packages/engine/src/python/replicate-db.ts
import { python } from '@trigger.dev/python'
import { env } from '@workspace/env'
import { pyArgs, pyPath, pyStreamingResult } from './utils'

type ReplicateDbArgs = {
  workspace_id: string
  source_id: string
  dialect?: string
  file_path?: string
  tables_metadata?: {
    name: string
    columns: {
      name: string
    }[]
  }[]
}

type ReplicateDbEnvVars = {
  host?: string
  port?: string
  user?: string
  password?: string
  database?: string
  schema?: string
}

type ReplicateDbResult = {
  status: 'success' | 'error'
  message: string
}

export async function replicateDb(
  args: ReplicateDbArgs,
  { envVars }: { envVars?: ReplicateDbEnvVars } = {},
): Promise<ReplicateDbResult> {
  const streamingResult = python.stream.runScript(
    pyPath('replicate_db'),
    pyArgs(args),
    {
      env: {
        PYTHON_ENV: env.NODE_ENV,
        DATABASE_CLIENT_HOST: envVars?.host,
        DATABASE_CLIENT_PORT: envVars?.port,
        DATABASE_CLIENT_USER: envVars?.user,
        DATABASE_CLIENT_PASSWORD: envVars?.password,
        DATABASE_CLIENT_DATABASE: envVars?.database,
        DATABASE_CLIENT_SCHEMA: envVars?.schema,
      },
    },
  )

  const result = await pyStreamingResult<ReplicateDbResult>(streamingResult)

  return result
}
